Il2CppInterop.Runtime/Runtime/Il2CppObjectPool.cs

```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Runtime.CompilerServices;
+using Il2CppInterop.Runtime.InteropTypes;
+
+namespace Il2CppInterop.Runtime.Runtime;
+
+public static class Il2CppObjectPool
+{
+    private static readonly ConcurrentDictionary<IntPtr, WeakReference<Il2CppObjectBase>> s_cache = new();
+
+    internal static void Remove(IntPtr ptr)
+    {
+        s_cache.TryRemove(ptr, out _);
+    }
+
+    public static T Get<T>(IntPtr ptr)
+    {
+        var ownClass = IL2CPP.il2cpp_object_get_class(ptr);
+        if (RuntimeSpecificsStore.IsInjected(ownClass))
+        {
+            var monoObject = ClassInjectorBase.GetMonoObjectFromIl2CppPointer(ptr);
+            if (monoObject is T monoObjectT) return monoObjectT;
+        }
+
+        if (s_cache.TryGetValue(ptr, out var reference) && reference.TryGetTarget(out var cachedObject))
+        {
+            if (cachedObject is T cachedObjectT) return cachedObjectT;
+            // This leaves the case when you cast to an interface handled as if nothing was cached
```

Comment:
Shouldn't we clear the `pooledPtr` of the `cachedObject` here so it won't remove the newer object on finalize?
Also, the question is do we really want this behavior? Would it hurt the performance too much to also differentiate the s_cache by type? (we can't use generics because there is no easy way to get T in the `~Il2CppObjectBase`)

Reply:
I'm not sure it's worth the added complexity

Reply:
I changed the pooledPtr thing.

In terms of differentiating the cache by type, I do not think the added complexity is warrented for such a minimal benefit